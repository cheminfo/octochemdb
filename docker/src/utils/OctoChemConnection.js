import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
