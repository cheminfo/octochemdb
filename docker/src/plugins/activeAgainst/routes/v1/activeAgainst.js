import { OctoChemConnection } from '../../../../server/utils.js';
import debugLibrary from '../../../../utils/Debug.js';

const debug = debugLibrary('activeAgainst');

// export default activeAgainst;
const activeAgainst = {
  method: 'GET',
  schema: {
    summary: 'Retrieve all entries from the activeAgainst collection',
    description:
      'This route retrieves all entries from the activeAgainst collection, which contains information about the taxonomy of the targets organisms to which the active compounds are active. \n This is could be useful to determine the kind of bioassays that are performed (e.g. 20% of the bioassays are performed against bacteria, etc.) or to know if exists active compounds against a specific organism.',
  },
  handler: searchHandler,
};
export default activeAgainst;
/**
 * @description Get all entries from the activeAgainst collection
 * @return {Promise} Returns statistics about the collection
 */

async function searchHandler() {
  let connection;
  try {
    // Get collections from the database
    connection = new OctoChemConnection();
    const collection = await connection.getCollection('activeAgainst');
    // Get all entries from the activeAgainst collection
    const results = await collection
      .aggregate([{ $match: { _id: { $exists: true } } }])
      .toArray();
    return { data: results };
  } catch (e) {
    if (connection) {
      debug.fatal(e.message, {
        collection: 'activeAgainst',
        connection,
        stack: e.stack,
      });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    debug.trace('Closing connection');
    if (connection) await connection.close();
  }
}
