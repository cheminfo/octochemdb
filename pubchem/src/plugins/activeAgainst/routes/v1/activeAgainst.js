import { PubChemConnection } from '../../../../server/utils.js';
import Debug from '../../../../utils/Debug.js';

const debug = Debug('activeAgainst');

// export default activeAgainst;
const activeAgainst = {
  method: 'GET',
  schema: {
    summary: 'Retrieve all entries from the activeAgainst collection',
  },
  handler: searchHandler,
};
export default activeAgainst;

/**
 * @description Get all entries from the activeAgainst collection
 * @return {Promise} Returns statistics about the collection
 */

async function searchHandler() {
  let connection;
  try {
    // Get collections from the database
    connection = new PubChemConnection();
    const collection = await connection.getCollection('activeAgainst');
    // Get all entries from the activeAgainst collection
    const results = await collection
      .aggregate([{ $match: { _id: { $exists: true } } }])
      .toArray();
    return { data: results };
  } catch (e) {
    if (connection) {
      debug(e, { collection: 'activeAgainst', connection });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    debug('Closing connection');
    if (connection) await connection.close();
  }
}
