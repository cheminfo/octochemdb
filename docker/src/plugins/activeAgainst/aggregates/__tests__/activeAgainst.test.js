import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../utils/OctoChemConnection.js';
import activeAgainst from '../../routes/v1/activeAgainst.js';

test('getAll (activeAgainst)', async () => {
  const connection = new OctoChemConnection();

  while (true) {
    const collection = await connection.getCollection('activeAgainst');
    if ((await collection.countDocuments()) === 1) {
      break;
    }
  }

  const results = await activeAgainst.handler();
  expect(results.data).toMatchSnapshot();
});
