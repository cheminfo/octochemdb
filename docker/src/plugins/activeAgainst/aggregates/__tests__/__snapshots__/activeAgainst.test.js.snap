// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAll (activeAgainst) 1`] = `
[
  {
    "_id": "ArchaeaCandidatusBorrarchaeotaCandidatusBorrarchaeiaCandidatusBorrarchaealesCandidatusBorrarchaeaceaeCandidatusBorrarchaeum",
    "count": 20,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
]
`;

exports[`id search (activesOrNaturals) 1`] = `
[
  {
    "_id": "ArchaeaCandidatusBorrarchaeotaCandidatusBorrarchaeiaCandidatusBorrarchaealesCandidatusBorrarchaeaceaeCandidatusBorrarchaeum",
    "count": 20,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
]
`;
